tests/benchmarks/test_model_schema_generation.py

```diff
@@ -219,3 +238,104 @@ def serialize_model(self) -> Any:
             return self.field
 
     benchmark(rebuild_model, ModelWithValidator)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_str_discriminator_schema_generation(benchmark):
+    class Cat(BaseModel):
+        pet_type: Literal['cat']
+        meows: int
+
+    class Dog(BaseModel):
+        pet_type: Literal['dog']
+        barks: float
+
+    class Lizard(BaseModel):
+        pet_type: Literal['reptile', 'lizard']
+        scales: bool
+
+    class Model(BaseModel):
+        pet: Union[Cat, Dog, Lizard] = Field(..., discriminator='pet_type')
+        n: int
+
+    benchmark(rebuild_model, Model)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_callable_discriminator_schema_generation(benchmark):
+    class Pie(BaseModel):
+        time_to_cook: int
+        num_ingredients: int
+
+    class ApplePie(Pie):
+        fruit: Literal['apple'] = 'apple'
+
+    class PumpkinPie(Pie):
+        filling: Literal['pumpkin'] = 'pumpkin'
+
+    def get_discriminator_value(v: Any) -> str:
+        if isinstance(v, dict):
+            return v.get('fruit', v.get('filling'))
+        return getattr(v, 'fruit', getattr(v, 'filling', None))
+
+    class ThanksgivingDinner(BaseModel):
+        dessert: Annotated[
+            Union[
+                Annotated[ApplePie, Tag('apple')],
+                Annotated[PumpkinPie, Tag('pumpkin')],
+            ],
+            Discriminator(get_discriminator_value),
+        ]
+
+    benchmark(rebuild_model, ThanksgivingDinner)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_callable_discriminator_nested_schema_generation(benchmark):
```

Comment:
I don't think it really matters to have this benchmark. The goal of having benchmarks with discriminators was to cover most of the `Discriminator._convert_schema` code path, and more importantly the `apply_discriminators` cleaning step.

Here, having a nested union doesn't really relate to it, and is probably duplicating benchmarks of "normal" unions.

Reply:
Thanks for the insight. I'm always happy to learn a bit more about pydantic's internals. AI will remove them shortly